spring-integration-smb/src/main/java/org/springframework/integration/smb/outbound/SmbOutboundGateway.java

```diff
@@ -134,4 +187,88 @@ protected SmbFile enhanceNameWithSubDirectory(SmbFile file, String directory) {
 		}
 	}
 
+	@Override
+	protected List<?> ls(Message<?> message, Session<SmbFile> session, String dir) throws IOException {
+		return doTask(message, session, () -> super.ls(message, session, dir));
+	}
+
+	@Override
+	protected List<String> nlst(Message<?> message, Session<SmbFile> session, String dir) throws IOException {
+		return doTask(message, session, () -> super.nlst(message, session, dir));
+	}
+
+	@Override
+	protected File get(Message<?> message, Session<SmbFile> session, String remoteDir, String remoteFilePath,
+			String remoteFilename, SmbFile fileInfoParam) throws IOException {
+
+		return doTask(message, session,
+				() -> super.get(message, session, remoteDir, remoteFilePath, remoteFilename, fileInfoParam));
+	}
+
+	@Override
+	protected List<File> mGet(Message<?> message, Session<SmbFile> session, String remoteDirectory,
+			String remoteFilename) throws IOException {
+
+		return doTask(message, session,
+				() -> super.mGet(message, session, remoteDirectory, remoteFilename));
+	}
+
+	@Override
+	protected boolean rm(Message<?> message, Session<SmbFile> session, String remoteFilePath) throws IOException {
+		return doTask(message, session, () -> super.rm(message, session, remoteFilePath));
+	}
+
+	@Override
+	protected boolean mv(Message<?> message, Session<SmbFile> session, String remoteFilePath, String remoteFileNewPath)
+			throws IOException {
+
+		return doTask(message, session,
+				() -> super.mv(message, session, remoteFilePath, remoteFileNewPath));
+	}
+
+	@Override
+	protected String put(Message<?> message, Session<SmbFile> session, String subDirectory) {
+		try {
+			return doTask(message, session,
+					() -> super.put(message, session, subDirectory));
+		}
+		catch (IOException e) {
+			throw new UncheckedIOException(e);
+		}
+	}
+
+	@Override
+	protected List<String> mPut(Message<?> message, Session<SmbFile> session, File localDir) {
+		try {
+			return doTask(message, session,
+					() -> super.mPut(message, session, localDir));
+		}
+		catch (IOException e) {
+			throw new UncheckedIOException(e);
+		}
+	}
+
+	private <V> V doTask(Message<?> message, Session<SmbFile> session, Callable<V> task)
+			throws IOException {
+
+		try {
+			return task.call();
```

Comment:
Doesn't look like there is an extra work in this `doTask()`: you just do plain delegation to the provided task and therefore you call that `super.` and that's.
From here it looks like all these new method over here are without value and we simply can rely on the impls in the super class.

What do I miss, please?

Reply:
This code is based/follows the same logic in FtpOutboundGateway, and is a refactored version of 'doInWorkingDirectory' private method.

There is no concept of a working directory in the SMB client code/dependent library like what we have in 'FTPClient'. This method simply wraps the 'task.call()' in a safe try/catch.

All this logic was in the 'FtpOutboundGateway' class which I originally stripped out during the port to SMB. I have since found out that they are useful methods and will work with the SMB client code and dependent libraries.

Reply:
I'll refactor it if it doesn't make sense for this new gateway... and remove 'doTask'. it is somewhat redundant. looks like the code needs cleanup anyways...